### PR TITLE
[TRTLLM-15160][feat] Add post-attention o_proj gate hook to shared MLA base and make trtllm::kda_decode inplace-only

### PR DESCRIPTION
**Pull request by WeiHaocheng**
Description

  Two independent, low-risk PyTorch-backend refactors.

  1. Make trtllm::kda_decode inplace-only — Drop the optional output tensor from the op schema so the CUDA kernel only ever writes into a caller-provided buffer.
  Allocation moves to the Python wrapper, and the KDA mixer's optimized decode path supplies a persistent, pool-slot-sized, never-reallocated output buffer
  (mirroring the existing _cs_dense CUDA-graph contract). The fake registration is updated in lockstep and the op's mutated tensors are registered in inplace_info
  (alias-declaration order). No behavior change today (KDA decode runs eagerly under CUDA graphs) — correct prep for torch.compile.

  2. Add a post-attention o_proj gate hook to the shared MLA base — Introduce an overridable _apply_output_gate(hidden_states, attn_output) on the base MLA, applied
  to o_proj's input in MLA.forward before _project_output. Since o_proj is never fused into the register_to_config custom op, the hook runs identically for both
  the custom-op and eager branches, and before the helix-CP reduce-scatter. The base default is identity, so the DeepSeek/Eagle subclasses are unchanged. Kimi K3
  MLA overrides the hook with its sigmoid output gate and drops its forward() override, also picking up the base o_proj / helix-CP output-projection path.

  Validation

  - KDA decode unit tests (test_kda_decode.py, test_kda_decode_op.py): 77 passed on GB300 (SM103).
  - Shared MLA base change is a strict no-op for the DeepSeek family — test_attention_mla.py gives byte-identical results with and without the change.
  - Kimi K3 model-level parity: GSM8K DEP16/TP16 = 97.04, matching the ~96.8 reference within noise (no regression from the o_proj-gate refactor).

  Notes

  No API-signature changes and no new LLM args. The two changes are independent concerns and can be split into two PRs if preferred.

## Summary

- Made `trtllm::kda_decode` strictly in-place.
- Moved output allocation to the Python wrapper.
- Added mutation metadata for compilation and updated fake registration.
- Added a persistent KDA decode output buffer for CUDA graph reuse.
- Added the `_apply_output_gate` hook to shared `MLA`.
- Moved Kimi K3 sigmoid gating into the hook.
- Preserved identity behavior for DeepSeek and Eagle MLA.
- Updated KDA decode and MLA hook tests.

## Dev Engineer Review

- The C++, Python wrapper, fake implementation, and mutation metadata use consistent in-place semantics.
- The wrapper validates caller-provided output tensors and returns the populated buffer.
- The persistent KDA buffer supports optimized decode and CUDA graph pool sizing.
- The MLA hook runs before output projection and Helix reduce-scatter.
- The default hook preserves existing DeepSeek and Eagle behavior.
- No configuration files or test-list files changed.
- Reported parity results cover KDA decode, DeepSeek-family outputs, and Kimi K3 GSM8K.
- No additional correctness or regression issues are identified from the provided changes.

## QA Engineer Review

- Modified tests:
  - `tests/unittest/_torch/thop/parallel/test_kda_decode.py`
  - `tests/unittest/_torch/modules/test_mla_registry.py`
- `test_kda_decode` now passes an explicit output buffer and validates the in-place result.
- MLA registry tests cover identity gating, custom gating order, and sparse MLA rejection.
- No corresponding `test-db/` or `qa/` coverage entries are identified in the provided changes.
- Verdict: needs follow-up.

### CODE REVIEW

**Review by dc3671 — APPROVED** [review]

**Review by mikeiovine — APPROVED** [review]
Stamp on behalf of runtime devs, delegating proper review to @NVIDIA/trt-llm-kernels-devs; please ping me if you think this is not accurate

File: tensorrt_llm/_torch/modules/mla.py [hunk 1]
```diff
@@ -1833,6 +1854,12 @@ def forward(
                 latent_cache_gen=latent_cache_gen,
             )
 
+        # Post-attention, pre-o_proj hook (identity in the base; Kimi K3 MLA
```
Comment (pengbowang-nv) [inline tensorrt_llm/_torch/modules/mla.py]: NIT, non-blocking: Please simplify some of these comments, following a concise principle if you gets time. And note that we want comments to explain what the code do, not why the code is added for which specific model or bug. Thanks!

**Review by pengbowang-nv — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/modules/mla.py [hunk 2]
```diff
@@ -1833,6 +1854,12 @@ def forward(
...
+        # inserts its sigmoid output gate). Applied to o_proj's input for both
+        # branches -- o_proj is never fused into the custom op, so this is
+        # custom-op compatible -- and before _project_output's helix reduce-
+        # scatter.
+        attn_output[0] = self._apply_output_gate(output_hidden_states, attn_output[0])
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/modules/mla.py:1872]: The hook's documented contract ("runs on `o_proj`'s input tensor", "right before the row-parallel `o_proj`") only holds on the dense path. When `sparse_attn_hooks` is installed, `_project_output` dispatches to the hook instead, and for DeepSeek-V4 `attn_output[0]` can be a 3D post-`o_a_proj` O-LoRA intermediate (`attention_backend/sparse/deepseek_v4/module.py:295`) — different rank, different shape, different sharding — and that module has `o_a_proj`/`o_b_proj` rather than a single `o_proj`. The gate is also applied only to element 0, ignoring the extra buffers sparse backends append via `_create_outputs`.

Harmless today since the base is identity and K3 has no sparse hooks, but the docstring is the contract a future subclass will code against. Either scope it explicitly ("dense path only; undefined when `sparse_attn_hooks` is set") or add an assert in the base that a non-identity override and `sparse_attn_hooks is not None` don't coexist.

File: tensorrt_llm/_torch/modules/kimi_kda/kimi_kda_mixer.py [hunk 3]
```diff
@@ -584,12 +589,25 @@ def forward_decode(
                 device=x2d.device,
             )
             self._cs_dense = buf
+            # Same one-shot, pool-slot-sized, never-reallocated contract as the
+            # conv staging buffer above: the inplace-only kda_decode op writes
+            # its [B, 1, H, hd] result here, and captured graphs bind this
+            # pointer.
+            self._o_dense = torch.empty(
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/modules/kimi_kda/kimi_kda_mixer.py:574]: `_o_dense` is allocated only inside the `_cs_dense is None` branch, so the two buffers' lifetimes are implicitly coupled and the `else` branch has to defend against `_o_dense is None` with an assert that can only fire if that coupling ever breaks. Giving `_o_dense` its own `if self._o_dense is None:` guard (with the same `is_current_stream_capturing()` fallback) makes each buffer self-contained and lets the assert drop to a plain `self._o_dense.shape[0] >= B` check.

Minor, but note both asserts vanish under `python -O`, at which point `self._o_dense[:B]` on line 669 would raise `TypeError` instead of the intended message.

**Review by brnguyen2 — APPROVED** [review]
Approving — the comments below are optional touch-ups, not blockers.

The KDA half holds up: I checked the new `inplace_info` keys against the other `-> ()` ops already in that table (`logits_bitmask`, `inplace_slice_copy`, both keyed from 1), and the `a!`..`e!` declaration order in the schema matches the 1..5 mapping. The Fp4 assert hoisted out of the `register_to_config` branch is also not a new failure mode — `create_output` on the eager path already asserted the same invariant.

Two things:

**Split this into two PRs.** The repo's guidance is one concern per PR, and these are genuinely independent: a C++ op-schema change plus compilation metadata, and a model-behavior refactor across MLA/Kimi K3. They'll pull different reviewers, and the commits are already separate so the split is nearly free.

**The MLA hook has no unit coverage.** K3's gate correctness now depends entirely on the base calling `_apply_output_gate` between attention and `_project_output` — a contract enforced only by an end-to-end GSM8K number that nobody will re-run when `MLA.forward` is next refactored. A small test with a stub `MLA` subclass whose hook returns a sentinel (asserting it lands on `_project_output`'s input, on both the `register_to_config` and eager branches) would pin the ordering cheaply. Byte-identical `test_attention_mla.py` results only prove the base default is identity, which is the easy half.

Also worth calling out in the description: K3 dropping its `forward()` override means its output projection now goes through `_helix_cp_output_projection` and passes `layer_idx` to `o_proj`, where it previously called `self.o_proj(attn_out)` bare. That's presumably the intent, but it's a behavior delta the summary reads as a pure no-op.

**Review by kris1025 — APPROVED** [review]
